simpleParser.go

```diff
@@ -0,0 +1,470 @@
+	for strings.HasSuffix(ctx.tokens[ctx.currentTokenIndex], ")") {
+		ctx.tokens = append(ctx.tokens, "")
+		copy(ctx.tokens[ctx.currentTokenIndex+1:], ctx.tokens[ctx.currentTokenIndex:])
+		ctx.tokens[ctx.currentTokenIndex] = strings.TrimSuffix(ctx.tokens[ctx.currentTokenIndex], ")")
+		ctx.tokens[ctx.currentTokenIndex+1] = ")"
+	}
+	return nil
+}
+
+func (ctx *expressionParserContext) handleCloseParenBookKeeping() error {
+	if ctx.parenDepth == 0 {
+		return ErrorParenMismatch
+	}
+	ctx.parenDepth--
+	return nil
+}
+
+func (ctx *expressionParserContext) getCurrentToken() (string, ParseTokenType, error) {
+	if ctx.currentTokenIndex >= len(ctx.tokens) {
+		return "", TokenTypeInvalid, ErrorNoMoreTokens
+	}
+	token := ctx.tokens[ctx.currentTokenIndex]
+
+	// Values by def should be enclosed within single quotes
+	valueRegex := regexp.MustCompile(`^\'.*\'$`)
+	// Or Values can be int or floats by themselves (w/o alpha char)
+	valueNumRegex := regexp.MustCompile(`^(-?)(0|([1-9][0-9]*))(\\.[0-9]+)?$`)
+
+	if tokenIsOpType(token) {
+		return token, TokenTypeOperator, nil
+	} else if valueRegex.MatchString(token) {
+		return token, TokenTypeValue, nil
+	} else if valueNumRegex.MatchString(token) {
+		return token, TokenTypeValue, nil
+	} else if token == "true" {
+		return token, TokenTypeTrue, nil
+	} else if token == "false" {
+		return token, TokenTypeFalse, nil
+	} else if token != "(" && strings.HasPrefix(token, "(") {
+		ctx.handleOpenParen()
+		token = ctx.tokens[ctx.currentTokenIndex]
+		return token, TokenTypeParen, nil
+	} else if token != ")" && strings.HasSuffix(token, ")") {
+		ctx.handleCloseParen()
+		return ctx.getCurrentToken()
+	} else if token == ")" {
+		return token, TokenTypeEndParen, ctx.handleCloseParenBookKeeping()
+	} else {
+		return token, TokenTypeField, nil
+	}
+}
+
+func (ctx *expressionParserContext) setupForToken(tokenType ParseTokenType) error {
+	if ctx.parserTree.NumNodes() == 0 {
+		// First time setup already done in constructor
+		return nil
+	}
+	switch tokenType {
+	case TokenTypeParen:
+		// TODO - Need special handling
+	}
+	return nil
+}
+
+func (ctx *expressionParserContext) checkTokenTypeWithinContext(tokenType ParseTokenType) error {
+	switch ctx.subCtx.currentMode {
+	case fieldMode:
+		if tokenType == TokenTypeParen {
+			return ErrorNeedToStartNewCtx
+		} else if tokenType != TokenTypeField && tokenType != TokenTypeTrue && tokenType != TokenTypeFalse {
+			return fmt.Errorf("Error: For field mode, token must be field type - received %v", tokenType.String())
+		}
+	case opMode:
+		if tokenType == TokenTypeEndParen {
+			ctx.skipAdvanceMode = true
+			return NonErrorOneLayerDone
+		} else if tokenType != TokenTypeOperator {
+			return fmt.Errorf("Error: For operator mode, token must be operator type - received %v", tokenType.String())
+		}
+	case valueMode:
+		if tokenType == TokenTypeParen {
+			return ErrorNeedToStartNewCtx
+		} else if tokenType == TokenTypeField {
+			return ErrorNeedToStartOneNewCtx
```

Comment:
i am not following the logic here.
is this for expressions like "field1 > field2", where the another field shows up in the position normally for values? is it that you are creating a subcontext just for field2? 
if so, the subcontext should return after it completes processing a field. as of now it seems that it is returning after it processes a value in advanceToken().

Reply:
First, the code has been rewritten in the newer commit. See: https://github.com/nelio2k/gojsonsm/blob/94407c0854857b0b2157331d1ea503f2e63ccab4/simpleParser.go#L445
To answer your question: we don't technically allow field1 op field2. So this is not the case here.
The idea here is to address weird syntax (but valid) like:
```
true && field1 op value1
```
Other than that, I don't think we'll usually hit it.